feat: Secrets Manager plugin

crystall-bitquill opened the pull request:
### Summary

<!--- General summary / title -->
Add Secrets Manager plugin

### Description

<!--- Details of what you changed -->
- Added AWS sdk Secrets Manager dependency
- Added factory class and plugin class for Secrets Manager plugin
- Added error message for being unable to retrieve secrets
- Added region and secret ID properties for plugin
- Plugin class fetches secret based on region and secret ID properties during `openInitialConnection` function and passes updated credentials to the next plugin

### Additional Reviewers

<!-- Any additional reviewers -->

ColinKYuen committed:
feat: Failover Connection Plugin (#174)

- Failover functionality refactored into a connection plugin.
- Upstream changes from MySQL 8.0.28 community driver.
- Docker containers are now created as part of testing and used to run both unit and integration tests.
- Failover performance metrics tracked based on a cluster. With options to enable additional performance metrics per instance.
- Fixed issue in Enhanced Failure Monitoring connection status check.

hsuamz committed:
docs: Updating changelog for 1.0.0 release

- Changelog date has been updated to match the release timeframe.
- Added description about integration tests for Hikari connection pool.

susanmdouglas-aws committed:
docs: Website update that removes Public Preview wording (#171)

This update removes comments noting that the project is a Public Preview, and updates the Copyright year on the web page to 2022.  It should not be distributed until the project is no longer considered a public preview.

Signed-off-by: susanmdouglas <susandou@amazon.com>

susanmdouglas-aws committed:
docs: This update to the README.md file bumps the installation version in the README.md file to 1.0.0 and removes the 'Preview' note (#173)

This update to the README.md file bumps the installation version in the README.md file to 1.0.0 and removes the 'Preview' note.

Signed-off-by: susanmdouglas <susandou@amazon.com>

ColinKYuen committed:
chore: Remove mentions of Public Preview (#176)

Remove mentions of Public Preview from Maven publishes & workflows for GitHub draft/release

hsuamz committed:
docs: Updating README.md Terminology

- Updating terminology to use allowlisting.

sergiyvamz committed:
upgrade to AWS SDK v.2.17.165 (#187)

Co-authored-by: sergiyv-bitquill <sergiyv@bitquilltech.com>

sergiyvamz committed:
fix concurrency issues with topology cache (#189)

juanpabloprado committed:
gradle 7.4.2 (#194)

sergiyvamz committed:
fix user documentation concerning default plugins and use of clusterI… (#198)

* fix user documentation concerning default plugins and use of clusterInstanceHostPattern parameter

sergiyvamz committed:
add snapshot builds (#199)

sergiyvamz committed:
increment subminor version for snapshot builds (#201)

Commit:
feat: add secrets manager plugin implementation
feat: fix property values, add error messages for secrets plugin
test: add constructor for Secrets Manager plugin unit tests

sergiyv-bitquill commented on src/main/core-api/java/com/mysql/cj/conf/PropertyDefinitions.java:
```diff
@@ -713,6 +713,16 @@ public enum DatabaseTerm {
+                // CATEGORY_AWS_SECRETS_MANAGER
+                //
+
+                new StringPropertyDefinition(PropertyKey.secretsManagerRegion, DEFAULT_VALUE_NULL_STRING, RUNTIME_MODIFIABLE,
```
Taking into account that Secret Manager plug will be relocated to a separate repo, there will be no access to the driver classes. So these new properties should be defined/maintained by plugin code.

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,121 @@
+    GetSecretValueResult getSecretValueResult;
+    String secret = "";
+    Properties updatedProperties = new Properties();
+    updatedProperties.putAll(connectionUrl.getOriginalProperties());
```
I think we can update original properties directly.

crystall-bitquill replied:
I don't think the ConnectionUrl class allows for editing/accessing the properties directly.

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,121 @@
+      throw new SQLException(Messages.getString("AWSSecretsManagerPlugin.1"));
+    }
+
+    ConnectionUrl newConnectionUrl = ConnectionUrl.getConnectionUrlInstance(connectionUrl.getDatabaseUrl(), updatedProperties);
```
I believe we can use the same object instance

crystall-bitquill replied:
I made a new ConnectionUrl instance because I wasn't able to change the connectionUrl properties directly.

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,121 @@
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+  private final PropertySet propertySet;
+  private final ObjectMapper mapper = new ObjectMapper();
```
It could be a local variable for openInitialConnection()

karenc-bq replied:
I prefer object mapper to be a final static variable @sergiyv-bitquill 
https://stackoverflow.com/questions/3907929/should-i-declare-jacksons-objectmapper-as-a-static-field

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,121 @@
+      }
+
+      JsonNode jsonObject = mapper.readTree(secret);
+      updatedProperties.put("username", jsonObject.get("username").asText());
```
Should we consider to cache obtained username and pass for some time like 15 mins (configurable by user)?

crystall-bitquill replied:
If the secrets are only fetched during openInitialConnection, is caching needed?

karenc-bq commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,121 @@
+    AWSSecretsManager client  = AWSSecretsManagerClientBuilder.standard()
+        .withRegion(this.region)
+        .build();
+    GetSecretValueRequest getSecretValueRequest = new GetSecretValueRequest().withSecretId(this.secretId);
```
```suggestion
    final GetSecretValueRequest getSecretValueRequest = new GetSecretValueRequest().withSecretId(this.secretId);
```

karenc-bq commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,121 @@
+    GetSecretValueRequest getSecretValueRequest = new GetSecretValueRequest().withSecretId(this.secretId);
+    GetSecretValueResult getSecretValueResult;
+    String secret = "";
+    Properties updatedProperties = new Properties();
```
```suggestion
    final Properties updatedProperties = new Properties();
```

karenc-bq commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,121 @@
+
+    try {
+      getSecretValueResult = client.getSecretValue(getSecretValueRequest);
+      if (getSecretValueResult.getSecretString() != null) {
+        secret = getSecretValueResult.getSecretString();
```
```suggestion
      final String secretString = getSecretValueResult.getSecretString();
      secret = secretString != null ? secretString : "";
```

karenc-bq commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,121 @@
+      }
+
+      JsonNode jsonObject = mapper.readTree(secret);
+      updatedProperties.put("username", jsonObject.get("username").asText());
```
https://stackoverflow.com/a/40134224
It's more preferable to use textValue()

sergiyv-bitquill commented on build.gradle.kts:
```diff
@@ -232,6 +232,7 @@ dependencies {
     testImplementation("com.zaxxer:HikariCP:4.0.3")
 
     implementation("com.amazonaws:aws-java-sdk-rds:1.12.128")
+    implementation("com.amazonaws:aws-java-sdk-secretsmanager:1.12.213")
```
This dependency should be runtime-only dependency. 
https://dzone.com/articles/optional-dependencies

karenc-bq commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,114 @@
+
+    try {
+      final String secret = secretCache.getSecretString(secretId);
+      final ObjectMapper mapper = new ObjectMapper();
```
Object mapper can be a static final variable

sergiyv-bitquill commented on src/main/resources/com/mysql/cj/LocalizedErrorMessages.properties:
```diff
@@ -1099,4 +1099,7 @@ ClusterAwareReaderFailoverHandler.2=[ClusterAwareReaderFailoverHandler] Connecte
\ No newline at end of file
+ClusterAwareReaderFailoverHandler.6=[ClusterAwareReaderFailoverHandler] {0} was called with an invalid (null or empty) topology
+
+AWSSecretsManagerPlugin.1=[AWSSecretsManagerPlugin] Required dependencies 'com.amazonaws.services.secretsmanager.AWSSecretsManager' and/or 'com.amazonaws.secretsmanager.caching.SecretCache' are not on the classpath
```
I'd replace exact class names with dependent library names

sergiyv-bitquill replied:
Similar comment about plug to be extracted to another repo, this resource file won't be available for modifications.

sergiyv-bitquill commented on src/main/core-api/java/com/mysql/cj/conf/PropertyKey.java:
```diff
@@ -301,6 +301,10 @@ public enum PropertyKey {
     monitorDisposalTime("monitorDisposalTime", true),
 
+    // AWS Secrets Manager Plugin
+    secretsManagerSecretId("secretsManagerSecretId", true),
```
We can't define new properties in this class. As soon as plugin class be extracted to another repo there won't be available for modifications. Such properties need to be defined in plugin class.

sergiyv-bitquill commented on build.gradle.kts:
```diff
@@ -237,6 +237,8 @@ dependencies {
     implementation("org.javassist:javassist:3.28.0-GA")
     implementation("org.slf4j:slf4j-api:1.7.30")
     implementation("com.oracle.oci.sdk:oci-java-sdk-common:2.13.0")
+    compileOnly("com.amazonaws:aws-java-sdk-secretsmanager:1.12.213")
```
These libraries belong to AWS SDK v.1.x, is there similar libraries for v.2.x? If they exist, we need to migrate to v.2.x

crystall-bitquill replied:
I was able to find something for the secrets manager, but not the secrets manager caching.

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,132 @@
+      // Attempt to open initial connection with provided credentials
+      this.nextPlugin.openInitialConnection(connectionUrl);
+    } catch (SQLException connectionFailedException) {
+
```
Need to check this SQLException is actually about failing opening a connection. Otherwise, the exception should be re-thrown to let caller code to handle it.

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,147 @@
+    } catch (SQLException connectionFailedException) {
+
+      // If the thrown error is because user access was denied, get new credentials
+      if (Objects.equals(connectionFailedException.getSQLState(), SQLSTATE_ACCESS_ERROR)) {
```
Why not to use a simple "=="?

karenc-bq replied:
https://stackoverflow.com/questions/513832/how-do-i-compare-strings-in-java
Objects.equals is preferred when comparing strings in Java

sergiyv-bitquill replied:
I see. Since SQLSTATE_ACCESS_ERROR is constant and it's never null, more readable code would be
SQLSTATE_ACCESS_ERROR.equals(connectionFailedException.getSQLState())

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,147 @@
+    } catch (SQLException connectionFailedException) {
+
+      // If the thrown error is because user access was denied, get new credentials
+      if (Objects.equals(connectionFailedException.getSQLState(), SQLSTATE_ACCESS_ERROR)) {
```
Why not to use a simple "=="?

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,147 @@
+import com.mysql.cj.log.Log;
+
+import com.amazonaws.secretsmanager.caching.SecretCache;
+import com.amazonaws.services.secretsmanager.AWSSecretsManagerClientBuilder;
```
com.amazonaws.* packages belong to AWS SDK v.1.x
Proper packages would be software.amazon.awssdk.*

sergiyv-bitquill replied:
com.amazonaws.* packages belong to AWS SDK v.1.x
Proper packages would be software.amazon.awssdk.*

karenc-bq committed:
fix: incorrect initial connection properties (#211)

karenc-bq committed:
fix: filter out methods that don't require failover (#212)

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,183 @@
+  public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
+    final Properties updatedProperties = new Properties();
+    updatedProperties.putAll(connectionUrl.getOriginalProperties());
+    String secret = secretCache.get(this.secretId);
```
Is secretId unique among all regions? Or secretId is unique within a single region? A key for LRU cache may need to include region as well.

sergiyv-bitquill replied:
How we handle a case where cache is empty and secret is null?

sergiyv-bitquill replied:
The cache stores no a parsed user name and password but a JSON string that contains them. That means that every time the plugin needs user name and password, it needs to parse JSON. It's a bit of wasting a time. Is it possible to store pre-parsed value in the cache?

crystall-bitquill replied:
If the cache is empty, the driver will attempt to get the secret. If there are any exceptions when getting that secret, including if the secret does not exist, the driver will throw an SQLException.

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,183 @@
+  private final Log logger;
+  private final String secretId;
+  private final Region region;
+  private final LRUCache<String, String> secretCache = new LRUCache<>(100);
```
static?

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,222 @@
+  private final GetSecretValueRequest getSecretValueRequest;
+
+  @JsonIgnoreProperties(ignoreUnknown = true)
+  static class Secret {
```
Should we extract it into a separate file?

karenc-bq replied:
I think this is fine here. I prefer putting it at the bottom tho.

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,222 @@
+    }
+  }
+
+  public AWSSecretsManagerPlugin(
```
I see two similar constructors. Can we optimize them in order to avoid code repetition?

karenc-bq replied:
I believe the second constructor is for testing purposes and should be package private.

sergiyv-bitquill commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,222 @@
+        this.nextPlugin.openInitialConnection(newConnectionUrl);
+
+      } else {
+        this.logger.logError(ERROR_INITIAL_CONN_FAILED);
```
I don't think we need to log here.

karenc-bq commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,222 @@
+
+  Secret getCurrentCredentials() throws SQLException {
+    try {
+      final GetSecretValueResponse valueResponse = this.secretsManagerClient.getSecretValue(this.getSecretValueRequest);
+      String secretString = valueResponse.secretString();
+      final ObjectMapper mapper = new ObjectMapper();
+      final JsonNode jsonObject = mapper.readTree(secretString);
+      return mapper.treeToValue(jsonObject, Secret.class);
```
```suggestion
      final GetSecretValueResponse valueResponse = this.secretsManagerClient.getSecretValue(this.getSecretValueRequest);
      final ObjectMapper mapper = new ObjectMapper();
      return mapper.readValue(valueResponse.secretString(), Secret.class);
```
Does this work?

karenc-bq commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,218 @@
+    }
+  }
+
+  private void updateConnectionProperties(Properties props, Secret secret) throws SQLException {
```
```suggestion
  private void updateConnectionProperties(Properties props, Secret secret) {
```

karenc-bq commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,218 @@
+        try {
+          // Get updated credentials and store the secret in the cache
+          secret = getCurrentCredentials();
+          SECRET_CACHE.put(secretKey, secret);
+          updateConnectionProperties(updatedProperties, secret);
```
These two lines should be after the catch clause, right above `newConnectionUrl = ...`

karenc-bq commented on src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:
```diff
@@ -0,0 +1,218 @@
+
+    try {
+      // Attempt to open initial connection with cached secret
+      updateConnectionProperties(updatedProperties, secret);
+      newConnectionUrl = ConnectionUrl.getConnectionUrlInstance(connectionUrl.getDatabaseUrl(), updatedProperties);
```
These two lines don't throw SQLException, so they can be taken out of the try catch clause

karenc-bq committed:
chore: add workflow for stale issues (#218)

Add a Github action that runs every week to mark or close stale issues.

karenc-bq committed:
chore: update issue templates and documentation (#219)

- Add issue templates.
- Add logging information and read-write splitting information to the README.

crystall-bitquill committed:
test: unit tests for Secrets Manager plugin (#131)

karenc-bq committed:
Merge remote-tracking branch 'upstream/main' into rds-827-secrets-manager-plugin

# Conflicts:
#	.github/workflows/pull_request.yml
#	README.md
#	build.gradle.kts

karenc-bq committed:
chore: add punctuation and known issues

karenc-bq committed:
fix: propagate non-network errors during failover

Commit:
docs: update CHANGELOG.md for 1.1.0 release

chore: add punctuation

karenc-bq committed:
refactor: create a helper method to check if an exception is caused by network errors

karenc-bq committed:
chore: add sample code for secrets manager plugin

karenc-bq committed:
chore: update SQLException thrown in unit tests
